tools/ec2recovery.py

```diff
@@ -1,280 +1,211 @@
+def list_snapshots(dryrun, client, instance):
+    snapshots = client.describe_snapshots(
+        DryRun=dryrun,
+        Filters=[{
+            'Name': 'description',
+            'Values': ["Automated,Backup,%s,*" %
+                       instance]}])['Snapshots']
+    snapshot_map = {}
+    for snapshot in snapshots:
+        start_date = snapshot['StartTime'].date()
+        snapshot_id = snapshot['SnapshotId']
+        volume_id = snapshot['VolumeId']
+        description = snapshot['Description']
+        if not start_date in snapshot_map:
+            snapshot_map[start_date] = []
+        for tag in snapshot['Tags']:
+            if 'DeviceName' in tag['Key']:
+                if tag['Value'] == '/dev/sda1':
+                    continue
+                snapshot_map[start_date].append([
+                    volume_id, tag['Value'],
+                    snapshot_id, description])
+    for snapshot in snapshot_map:
+        print "\nSnapshot Date: %s" % snapshot
+        for value in snapshot_map[snapshot]:
+            print "\t VolumeId: %s (%s), SnapshotId: %s, Description: %s" % (
+                value[0], value[1],
+                value[2], value[3])
 
 
 def create_tags(tags):
     result = []
-    for i in range(len(tags)):
-        key = tags[i]['Key']
-        value = tags[i]['Value']
-        if key == 'Name':
-            value = "%s-restored" % value
-        if key == 'RestoreDate' or 'aws:' in key:
-            continue
-        result.append({
-            'Key': key,
-            'Value': value})
+    if tags:
+        for i in range(len(tags.split(','))):
+            pairs = tags[i].split(':')
+            key = pairs[0]
+            value = pairs[1]
+            result.append({
+                'Key': key,
+                'Value': value})
     result.append({
         'Key': 'RestoreDate',
         'Value': datetime.now().strftime("%Y-%m-%d %H:%M:%S")})
     return result
 
 
-def create_instance(session, dry_run, ami_id, instance, bdm, user_data):
-    # Create volumes from snapshots
-    ec2 = session.resource('ec2')
-    instances = ec2.create_instances(
-        DryRun=dry_run,
-        ImageId=ami_id,
-        MinCount=1,
-        MaxCount=1,
-        KeyName=instance['KeyName'],
-        SecurityGroupIds=pull_sg_ids(instance),
-        InstanceType=instance['InstanceType'],
-        SubnetId=instance['SubnetId'],
-        BlockDeviceMappings=bdm,
-        IamInstanceProfile={'Arn': instance['IamInstanceProfile']['Arn']},
-        UserData=user_data)
-    instance_id = instances[0].instance_id
-    restored = ec2.Instance(instance_id)
-    print "\n\tRestoring instance from snapshot..."
-
-    client = session.client('ec2')
-    waiter = client.get_waiter('instance_running')
-    waiter.wait(
-        DryRun=dry_run,
-        InstanceIds=[instance_id]
-    )
-    restored.create_tags(
-        DryRun=dry_run,
-        Tags=create_tags(instance['Tags']))
-    return instance_id
+def read_user_date(userdata):
```

Comment:
its great that you covered for completeness and more importantly made it optional, but afaics userdata is initialization, and on restore initialization is moot and generally untested and not idempotent, fine with leaving it in, but maybe print a warning if specified..